src/mixins/java/org/spongepowered/common/mixin/core/world/food/FoodDataMixin.java

```diff
@@ -0,0 +1,121 @@
+
+@Mixin(FoodData.class)
+public abstract class FoodDataMixin implements FoodDataBridge {
+
+    // @formatter: off
+    @Shadow private int foodLevel;
+    @Shadow private float saturationLevel;
+    @Shadow private float exhaustionLevel;
+    // @formatter: on
+
+    private Player impl$player;
+
+    @Override
+    public void bridge$setPlayer(Player player) {
+        this.impl$player = player;
+    }
+
+    @Redirect(method = "setFoodLevel", at = @At(value = "FIELD", target = "Lnet/minecraft/world/food/FoodData;foodLevel:I", opcode = Opcodes.PUTFIELD))
+    private void impl$setFoodLevel(FoodData self, int value) {
+        this.foodLevel = this.fireEventAndGetValue(Keys.FOOD_LEVEL, this.foodLevel, value);
+    }
+
+    @Redirect(method = "eat(IF)V", at = @At(value = "FIELD", target = "Lnet/minecraft/world/food/FoodData;foodLevel:I", opcode = Opcodes.PUTFIELD))
+    private void impl$eatSetFoodLevel(FoodData self, int value) {
+        this.foodLevel = this.fireEventAndGetValue(Keys.FOOD_LEVEL, this.foodLevel, value);
+    }
+
+    @Redirect(method = "tick", at = @At(value = "FIELD", target = "Lnet/minecraft/world/food/FoodData;foodLevel:I", opcode = Opcodes.PUTFIELD))
+    private void impl$tickDrainFoodLevel(FoodData self, int value) {
+        this.foodLevel = this.fireEventAndGetValue(Keys.FOOD_LEVEL, this.foodLevel, value);
+    }
+
+    @Redirect(method = "eat(IF)V", at = @At(value = "FIELD", target = "Lnet/minecraft/world/food/FoodData;saturationLevel:F", opcode = Opcodes.PUTFIELD))
+    private void impl$eatSetSaturationLevel(FoodData self, float value) {
+        this.saturationLevel = this.fireEventAndGetValue(Keys.SATURATION, (double) this.saturationLevel, (double) value).floatValue();
+    }
+
+    @Redirect(method = "tick", at = @At(value = "FIELD", target = "Lnet/minecraft/world/food/FoodData;saturationLevel:F", opcode = Opcodes.PUTFIELD))
+    private void impl$tickDrainSaturationLevel(FoodData self, float value) {
+        this.saturationLevel = this.fireEventAndGetValue(Keys.SATURATION, (double) this.saturationLevel, (double) value).floatValue();
+    }
+
+    @Redirect(method = "addExhaustion", at = @At(value = "FIELD", target = "Lnet/minecraft/world/food/FoodData;exhaustionLevel:F", opcode = Opcodes.PUTFIELD))
+    private void impl$addExhaustion(FoodData self, float value) {
+        this.exhaustionLevel = this.fireEventAndGetValue(Keys.EXHAUSTION, (double) this.exhaustionLevel, (double) value).floatValue();
+    }
+
+    @Redirect(method = "tick", at = @At(value = "FIELD", target = "Lnet/minecraft/world/food/FoodData;exhaustionLevel:F", opcode = Opcodes.PUTFIELD))
+    private void impl$tickDrainExhaustion(FoodData self, float value) {
+        this.exhaustionLevel = this.fireEventAndGetValue(Keys.EXHAUSTION, (double) this.exhaustionLevel, (double) value).floatValue();
+    }
+
+    private <E> E fireEventAndGetValue(Key<? extends Value<E>> key, E currentValue, E value) {
```

Comment:
The exhaustion one is spammy as that one is called every tick. I could drop that but implemented it for completeness. What do you think?

Reply:
No, don't drop it right this second, right now I'm just putting my thoughts out there. I imagine some of this can be mitigated in the ways I outline above though - if the value doesn't change, we don't need to fire the event.

Reply:
Sure, though it is modified every time player moves or interacts with the world. I myself don't see the exhaustion as necessary due the common case being for food related events to prevent it from draining, and the saturation is the last "visible" part that that the client sees. So in my eyes, cancelling saturation for that purpose is enough.

Reply:
> Sure, though it is modified every time player moves or interacts with the world

There are points in your injections where there might not be a change - `addExhaustion` may remain capped at 40 and it's possible on `tick` that `foodLevel` will remain at zero. Quite possible with `setFoodLevel` too.

> I myself don't see the exhaustion as necessary

But it is a data value and conceviably someone might want to monitor it.